diff -Naur Python-3.6.8-old/Lib/platform.py Python-3.6.8/Lib/platform.py
--- Python-3.6.8-old/Lib/platform.py    2018-12-23 21:37:14.000000000 +0000
+++ Python-3.6.8/Lib/platform.py        2021-05-01 17:12:48.557643873 +0000
@@ -297,7 +297,7 @@
 # and http://www.die.net/doc/linux/man/man1/lsb_release.1.html

 _supported_dists = (
-    'SuSE', 'debian', 'fedora', 'redhat', 'centos',
+    'SuSE', 'debian', 'fedora', 'redhat', 'centos', 'openela',
     'mandrake', 'mandriva', 'rocks', 'slackware', 'yellowdog', 'gentoo',
     'UnitedLinux', 'turbolinux', 'arch', 'mageia')

